lib/syskit/roby_app/remote_processes/server.rb

```diff
@@ -0,0 +1,408 @@
+                                Server.warn "   #{line}"
+                            end
+                        end
+
+                    elsif cmd_code == COMMAND_START
+                        name, deployment_name, name_mappings, options = Marshal.load(socket)
+                        options ||= Hash.new
+                        Server.debug "#{socket} requested startup of #{name} with #{options} and mappings #{name_mappings}"
+                        begin
+                            p = start_process(name, deployment_name, name_mappings, options)
+                            Server.debug "#{name}, from #{deployment_name}, is started (#{p.pid})"
+                            socket.write(RET_STARTED_PROCESS)
+                            Marshal.dump(p.pid, socket)
+                        rescue Interrupt
+                            raise
+                        rescue Exception => e
+                            Server.warn "failed to start #{name}: #{e.message}"
+                            (e.backtrace || Array.new).each do |line|
+                                Server.warn "   #{line}"
+                            end
+                            socket.write(RET_NO)
+                            socket.write Marshal.dump(e.message)
+                        end
+                    elsif cmd_code == COMMAND_END
+                        name, cleanup, hard = Marshal.load(socket)
+                        Server.debug "#{socket} requested end of #{name}"
+                        if (p = processes[name])
+                            begin
+                                end_process(p, cleanup: cleanup, hard: hard)
+                                socket.write(RET_YES)
+                            rescue Interrupt
+                                raise
+                            rescue Exception => e
+                                Server.warn "exception raised while calling #{p}#kill(false)"
+                                Server.log_pp(:warn, e)
+                                socket.write(RET_NO)
+                            end
+                        else
+                            Server.warn "no process named #{name} to end"
+                            socket.write(RET_NO)
+                        end
+                    elsif cmd_code == COMMAND_QUIT
+                        quit
+                    elsif cmd_code == COMMAND_UPLOAD_LOG
+                        begin
+                            host, port, certificate, user, password, localfile = Marshall.load(socket)
+                            Server.debug "#{socket} requested uploading a log file to FTP server"
+                            @active_threads << Thread.new do
+                                Thread.current.abort_on_exception = true
+                                upload_log(host, port, certificate, user, password, localfile)
+                                @active_threads.delete(Thread.current)
```

Comment:
That's a race condition. You can't modify an Array (which is not thread safe) in two threads at the same time. The very place where this is *guaranteed* to be bad is when you do `@active_threads.each(&:join)` <-- iterate and delete at the same time.

The better way would be to let the threads simply finish, and garbage-collect the terminated ones in the main `listen` loop, i.e.

~~~
@active_threads.delete_if do |thread|
   !thread.alive?
end
~~~